src/librustc/ty/wf.rs

```diff
@@ -157,33 +162,77 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
                 .collect()
     }
 
-    /// Pushes the obligations required for `trait_ref` to be WF into
-    /// `self.out`.
+    /// Pushes the obligations required for `trait_ref` to be WF into `self.out`.
     fn compute_trait_ref(&mut self, trait_ref: &ty::TraitRef<'tcx>, elaborate: Elaborate) {
+        let tcx = self.infcx.tcx;
         let obligations = self.nominal_obligations(trait_ref.def_id, trait_ref.substs);
 
         let cause = self.cause(traits::MiscObligation);
         let param_env = self.param_env;
 
         if let Elaborate::All = elaborate {
+            let trait_assoc_items = tcx.associated_items(trait_ref.def_id);
+
             let predicates = obligations.iter()
-                                        .map(|obligation| obligation.predicate.clone())
-                                        .collect();
-            let implied_obligations = traits::elaborate_predicates(self.infcx.tcx, predicates);
+                .map(|obligation| obligation.predicate.clone())
+                .collect();
+            let implied_obligations = traits::elaborate_predicates(tcx, predicates);
+            let item_span: Option<Span> = self.item.map(|i| i.span);
+            let item = &self.item;
             let implied_obligations = implied_obligations.map(|pred| {
-                traits::Obligation::new(cause.clone(), param_env, pred)
+                let mut cause = cause.clone();
+                match &pred {
+                    ty::Predicate::Projection(proj) => {
+                        if let Some(hir::ItemKind::Impl(.., impl_items)) = item.map(|i| &i.kind) {
+                            let trait_assoc_item = tcx.associated_item(proj.projection_def_id());
+                            if let Some(impl_item) = impl_items.iter().filter(|item| {
+                                item.ident == trait_assoc_item.ident
+                            }).next() {
+                                cause.span = impl_item.span;
+                                cause.code = traits::AssocTypeBound(
+                                    item_span,
+                                    trait_assoc_item.ident.span,
+                                );
+                            }
+                        }
+                    }
+                    ty::Predicate::Trait(proj) => {
+                    if let (
+                        ty::Projection(ty::ProjectionTy { item_def_id, .. }),
+                            Some(hir::ItemKind::Impl(.., impl_items)),
```

Comment:
this indentation looks surprising to me -- shouldn't this be aligned with the line before?